File: frontend/js/reports.js
// reports.js -- one tab per report; each report defines its API path,
// its own table columns, and how to format each column.
const REPORTS = [
    { key: "water-by-farmer", label: "Water by Farmer", path: "/reports/water-by-farmer",
      cols: [["farmer_id", "ID"], ["name", "Farmer"], ["total_water_used", "litres:Total Water Used"]] },
    { key: "water-by-canal", label: "Water by Canal", path: "/reports/water-by-canal",
      cols: [["canal_id", "ID"], ["canal_name", "Canal"], ["total_water_used", "litres:Total Water Used"]] },
    { key: "top-crops", label: "Highest Water-Consuming Crops", path: "/reports/top-crops",
      cols: [["crop_id", "ID"], ["crop_name", "Crop"], ["total_water_used", "litres:Total Water Used"]] },
    { key: "unpaid-farmers", label: "Farmers with Unpaid Bills", path: "/reports/unpaid-farmers",
      cols: [["farmer_id", "ID"], ["name", "Farmer"], ["phone", "Phone"], ["unpaid_bills", "Unpaid Bills"], ["total_due", "currency:Total Due"]] },
    { key: "daily-allocation", label: "Daily Water Allocation", path: "/reports/daily-allocation",
      cols: [["allocation_date", "date:Date"], ["schedule_count", "Schedules"], ["total_allocated", "litres:Total Allocated"]] },
    { key: "canal-utilization", label: "Canal Utilization", path: "/reports/canal-utilization",
      cols: [["canal_name", "Canal"], ["status", "status:Status"], ["total_schedules", "Total Schedules"], ["total_allocated_litres", "litres:Total Allocated"]] },
    { key: "usage-vs-scheduled", label: "Usage vs Scheduled", path: "/reports/usage-vs-scheduled",
      cols: [["schedule_id", "ID"], ["farmer_name", "Farmer"], ["scheduled_water", "litres:Scheduled"], ["actual_water", "litres:Actual"], ["difference", "litres:Difference"]] },
    { key: "dam-availability", label: "Current Dam Availability", path: "/reports/dam-availability",
      cols: [["water_level", "Level (m)"], ["available_water", "litres:Available Water"], ["release_rate", "litres:Release Rate"], ["recorded_at", "datetime:Recorded At"]] },
    { key: "complaints-by-type", label: "Complaints by Type", path: "/reports/complaints-by-type",
      cols: [["complaint_type", "Type"], ["total", "Total"]] },
    { key: "completed-schedules", label: "Completed Schedules Count", path: "/reports/completed-schedules",
      cols: [["completed_count", "Completed Schedules"]] },
];

let activeReport = null;

document.addEventListener("DOMContentLoaded", () => {
    const tabsEl = document.getElementById("report-tabs");
    tabsEl.innerHTML = REPORTS.map((r) => `<button class="btn btn-outline-primary btn-sm" data-key="${r.key}" onclick="loadReport('${r.key}')">${r.label}</button>`).join("");
    loadReport(REPORTS[0].key);
});

async function loadReport(key) {
    const report = REPORTS.find((r) => r.key === key);
    if (!report) return;
    activeReport = report;

    document.querySelectorAll("#report-tabs button").forEach((b) => b.classList.toggle("btn-primary", b.dataset.key === key));
    document.querySelectorAll("#report-tabs button").forEach((b) => b.classList.toggle("btn-outline-primary", b.dataset.key !== key));

    document.getElementById("report-title").textContent = report.label;
    document.getElementById("report-head").innerHTML = report.cols.map(([, spec]) => `<th>${spec.split(":").pop()}</th>`).join("");
    document.getElementById("report-body").innerHTML = `<tr><td colspan="${report.cols.length}">Loading...</td></tr>`;

    try {
        const res = await api.get(report.path);
        renderReportTable(res.data, report.cols);
    } catch (err) {
        showAlert("alert-box", err.message);
    }
}

function renderReportTable(rows, cols) {
    const tbody = document.getElementById("report-body");
    if (!rows.length) {
        tbody.innerHTML = `<tr><td colspan="${cols.length}">No data.</td></tr>`;
        return;
    }
    tbody.innerHTML = rows.map((row) => {
        const cells = cols.map(([field, spec]) => {
            const kind = spec.includes(":") ? spec.split(":")[0] : "";
            const v = row[field];
            let out = v;
            if (kind === "litres") out = formatLitres(v);
            else if (kind === "currency") out = formatCurrency(v);
            else if (kind === "date") out = formatDate(v);
            else if (kind === "datetime") out = formatDateTime(v);
            else if (kind === "status") out = badgeHtml(v);
            return `<td>${out === null || out === undefined ? "-" : out}</td>`;
        }).join("");
        return `<tr>${cells}</tr>`;
    }).join("");
}

File: backend/routes/reports.py
from flask import Blueprint
import mysql.connector
from db.connection import get_connection
from utils.responses import success_response, error_response

reports_bp = Blueprint("reports", __name__)


def _run(query, params=None):
    conn = None
    try:
        conn = get_connection()
        cur = conn.cursor(dictionary=True)
        cur.execute(query, params or ())
        return success_response(cur.fetchall())
    except (mysql.connector.Error, ConnectionError) as e:
        return error_response("Report query failed.", 500, e)
    finally:
        if conn:
            conn.close()


# Report 1: total water used by each farmer
@reports_bp.route("/water-by-farmer", methods=["GET"])
def water_by_farmer():
    return _run("""
        SELECT f.farmer_id, f.name, COALESCE(SUM(wu.measured_water),0) AS total_water_used
        FROM Farmers f
        LEFT JOIN Water_Schedules ws ON ws.farmer_id = f.farmer_id
        LEFT JOIN Water_Usage wu ON wu.schedule_id = ws.schedule_id
        GROUP BY f.farmer_id, f.name
        ORDER BY total_water_used DESC
    """)


# Report 2: total water used by each canal
@reports_bp.route("/water-by-canal", methods=["GET"])
def water_by_canal():
    return _run("""
        SELECT cn.canal_id, cn.canal_name, COALESCE(SUM(wu.measured_water),0) AS total_water_used
        FROM Canals cn
        LEFT JOIN Water_Schedules ws ON ws.canal_id = cn.canal_id
        LEFT JOIN Water_Usage wu ON wu.schedule_id = ws.schedule_id
        GROUP BY cn.canal_id, cn.canal_name
        ORDER BY total_water_used DESC
    """)


# Report 3: highest water-consuming crops
@reports_bp.route("/top-crops", methods=["GET"])
def top_crops():
    return _run("""
        SELECT c.crop_id, c.crop_name, COALESCE(SUM(wu.measured_water),0) AS total_water_used
        FROM Crops c
        JOIN Land_Parcels lp ON lp.crop_id = c.crop_id
        JOIN Water_Schedules ws ON ws.parcel_id = lp.parcel_id
        JOIN Water_Usage wu ON wu.schedule_id = ws.schedule_id
        GROUP BY c.crop_id, c.crop_name
        ORDER BY total_water_used DESC
    """)


# Report 4: farmers with unpaid bills
@reports_bp.route("/unpaid-farmers", methods=["GET"])
def unpaid_farmers():
    return _run("""
        SELECT f.farmer_id, f.name, f.phone, COUNT(b.bill_id) AS unpaid_bills, SUM(b.total_amount) AS total_due
        FROM Farmers f
        JOIN Bills b ON b.farmer_id = f.farmer_id
        WHERE b.status IN ('UNPAID', 'OVERDUE')
        GROUP BY f.farmer_id, f.name, f.phone
        HAVING total_due > 0
        ORDER BY total_due DESC
    """)


# Report 5: daily water allocation
@reports_bp.route("/daily-allocation", methods=["GET"])
def daily_allocation():
    return _run("""
        SELECT DATE(start_time) AS allocation_date, COUNT(*) AS schedule_count, SUM(approved_water) AS total_allocated
        FROM Water_Schedules
        WHERE status IN ('APPROVED','COMPLETED')
        GROUP BY DATE(start_time)
        ORDER BY allocation_date DESC
    """)


# Report 6: canal utilization
@reports_bp.route("/canal-utilization", methods=["GET"])
def canal_utilization_report():
    return _run("SELECT * FROM vw_canal_utilization ORDER BY total_allocated_litres DESC")


# Report 7: water usage vs scheduled allocation
@reports_bp.route("/usage-vs-scheduled", methods=["GET"])
def usage_vs_scheduled():
    return _run("""
        SELECT ws.schedule_id, f.name AS farmer_name, ws.approved_water AS scheduled_water,
               COALESCE(SUM(wu.measured_water),0) AS actual_water,
               (ws.approved_water - COALESCE(SUM(wu.measured_water),0)) AS difference
        FROM Water_Schedules ws
        JOIN Farmers f ON f.farmer_id = ws.farmer_id
        LEFT JOIN Water_Usage wu ON wu.schedule_id = ws.schedule_id
        WHERE ws.approved_water IS NOT NULL
        GROUP BY ws.schedule_id, f.name, ws.approved_water
        ORDER BY ws.schedule_id
    """)


# Report 8: current dam water availability
@reports_bp.route("/dam-availability", methods=["GET"])
def dam_availability():
    return _run("SELECT * FROM Dam_Status ORDER BY recorded_at DESC LIMIT 1")


# Report 9: number of complaints by type
@reports_bp.route("/complaints-by-type", methods=["GET"])
def complaints_by_type():
    return _run("SELECT complaint_type, COUNT(*) AS total FROM Complaints GROUP BY complaint_type ORDER BY total DESC")


# Report 10: number of completed irrigation schedules
@reports_bp.route("/completed-schedules", methods=["GET"])
def completed_schedules():
    return _run("SELECT COUNT(*) AS completed_count FROM Water_Schedules WHERE status = 'COMPLETED'")
